src/database/test/DatabaseTests.cpp

```diff
@@ -349,3 +355,430 @@ TEST_CASE("schema test", "[db]")
+        auto buyingStr = decoder::encode_b64(xdr::xdr_to_opaque(oe.buying));
+        double price = double(oe.price.n) / double(oe.price.d);
+
+        soci::transaction tx(session);
+
+        session << "INSERT INTO offers ( "
+                   "sellerid, offerid, sellingasset, buyingasset, "
+                   "amount, pricen, priced, price, flags, lastmodified "
+                   ") VALUES ( "
+                   ":v1, :v2, :v3, :v4, :v5, :v6, :v7, :v8, :v9, :v10 "
+                   ")",
+            soci::use(sellerIDStr), soci::use(oe.offerID),
+            soci::use(sellingStr), soci::use(buyingStr), soci::use(oe.amount),
+            soci::use(oe.price.n), soci::use(oe.price.d), soci::use(price),
+            soci::use(oe.flags),
+            soci::use(app.getLedgerManager().getLastClosedLedgerNum());
+
+        tx.commit();
+    };
+
+    auto prepOldSchemaDB = [addOneOldSchemaAccount, addOneOldSchemaTrustLine,
+                            addOneOldSchemaDataEntry,
+                            addOneOldSchemaOfferEntry](
+                               SchemaUpgradeTestApplication& app,
+                               AccOptLiabilitiesVec const& aols,
+                               TLOptLiabilitiesVec const& tlols,
+                               DataEntry const& de, OfferEntry const& oe) {
+        try
+        {
+            for (auto aol : aols)
+            {
+                addOneOldSchemaAccount(app, aol);
+            }
+        }
+        catch (std::exception& e)
+        {
+            CLOG(FATAL, "Database") << __func__ << ": exception " << e.what()
+                                    << " while adding old-schema accounts";
+            throw;
+        }
+
+        try
+        {
+            for (auto tlol : tlols)
+            {
+                addOneOldSchemaTrustLine(app, tlol);
+            }
+        }
+        catch (std::exception& e)
+        {
+            CLOG(FATAL, "Database") << __func__ << ": exception " << e.what()
+                                    << " while adding old-schema trustlines";
+            throw;
+        }
+
+        try
+        {
+            addOneOldSchemaDataEntry(app, de);
+        }
+        catch (std::exception& e)
+        {
+            CLOG(FATAL, "Database") << __func__ << ": exception " << e.what()
+                                    << " while adding old-schema data";
+            throw;
+        }
+
+        try
+        {
+            addOneOldSchemaOfferEntry(app, oe);
+        }
+        catch (std::exception& e)
+        {
+            CLOG(FATAL, "Database") << __func__ << ": exception " << e.what()
+                                    << " while adding old-schema offer";
+            throw;
+        }
+    };
+
+    auto testOneDBMode = [prepOldSchemaDB](Config::TestDbMode const dbMode) {
+        // A vector of optional Liabilities entries, for each of which the test
+        // will generate a valid account.
+        auto const accOptLiabilities = {
```

Comment:
As mentioned about the types above, I don't really understand why this approach is valuable. But regardless: why use different liabilities for accounts and trust lines, and why have repeats? Also you're missing the important (0, 0) case for accounts.

Reply:
Added (0,0) for accounts.  As for there being two different lists, I had two possibly-silly reasons:  I didn't want to give the impression that there was any dependency on their being the same, and I figured I'd take advantage of the there being two different types of ledger entries with liabilities by testing lists with different characteristics (different lengths, one beginning with a nullopt but not ending with one while the other was the other way around).  For the moment, having explained why, I'll leave it that way, but I'll squash the two lists into one if you still prefer I do so.